Sendmail plugin: Allow filtering for check-ins

pajowu opened the pull request:
This PR adds the ability to send mass emails to all attendees who have checked in (with the ability to filter for specific checkin lists). This is sort-of what was asked in #1369 (It is not possible to filter for users who never checked in)

raphaelm commented on src/pretix/plugins/sendmail/views.py:
```diff
@@ -183,6 +199,9 @@ def get_context_data(self, **kwargs):
                 itemcache[i['id']] for i in log.pdata.get('items', [])
             ]
+            log.pdata['checkin_lists'] = [
+                checkin_list_cache[i['id']] for i in log.pdata.get('checkin_lists', [])
```
You could maybe add a ``if i['id'] in checkin_list_cache`` to prevent crashes if a check-in list got deleted in the meantime

pajowu replied:
Will do

pajowu commented:
I considered having only one option first, similar to what there is currently for items. I decided against it because that would mean one of three things:
1. **Show all checkin lists, send the eMails to all chosen lists.**
This would mean that orders which were not checked in until now cannot be reached via mass email. This is not an option
2. **Show all checkin lists. Selecting none means sending to all users** (that match the other criteria)
I saw this as too unintuitiv
3. **Show a "not checked in" list**
This is probably the best idea but I decided against it first because it might a bit confusing.

Since 3. also solves the question in the PR better I will take a shot at it and update the PR once it is ready

raphaelm commented:
Hmm, I'm not sure I'm a fan of option two. I understand your reasoning and think either option two or the current approach is the best. I'll sleep on it ;)

pajowu commented:
Oops, I saw your message too late. The last commit adds a (slight hacky IMO) implementation of option 3. Depending on what you deem better can be merged. I *think* you can revert the commit if you want to merge option 2 (I enabled `Allow edits from maintainers.`), but I am also happy to revert it and add the small thing you asked for.

raphaelm commented:
Well, seeing you did that already, let's keep it that way for now and I'll have a look ;) Maybe I start do like it… Hopefully tomorrow, but I don't know, the day is rather full.

raphaelm commented on src/pretix/plugins/sendmail/tasks.py:
```diff
@@ -36,6 +39,9 @@ def send_mails(event: Event, user: int, subject: dict, message: dict, orders: li
                     continue
 
+                if forms.MailForm.NOT_CHECKED_IN not in checkin_lists and not any(str(c.list_id) in checkin_lists for c in p.checkins.all()):
+                    continue
```
The logic seems off here.

If I only select "Not checked in", it does send emails to everyone, even those who checked in, doesn't it?

pajowu commented:
I think reverting to the previous approach might also be a good Idea if more filters are added later (as proposed for example in #1313) to keep the form to a reasonable length.

pajowu commented:
I changed the input method to a checkbox for filtering at all which expands to a checkbox for filtering not-checked-in orders plus a ajax-based multi-select for filtering checkin lists.

![2019-10-20-114113_1595x184_scrot](https://user-images.githubusercontent.com/6545097/67157615-c58e6700-f32e-11e9-98ec-0fc0fde8ccea.png)

I also rebased the branch to the current master so it should be merge-able without problems.

raphaelm commented:
Thank you so much! :heart: 

I was planning for weeks to finish this one up, but didn't get around to. Now I felt bad for making you jump through even more hoops, so I did some more tweaks myself:

* I cleaned up some code that was changed through all the back-and-forth

* I resized the panel in the form so it looks better and does not "break up" the form

* I made a slight change in semantics of the options. With your version, it was *possible* to select both "only send to customers not checked in" *and* a check-in list, but the result of that operation was neither well-defined nor really useful. We could have just prevented that situation by blocking this combination, but I imagine that there is a real use case of "send to all people who are either not checked in at all or just checked in on list X" to exclude everyone who already checked in on a list Y. Therefore, I just removed the "Only" from the form fields and adjusted the logic accordingly.

* I fixed a "bug" in the query that existed before this PR as well, but turned worse through this PR. Previously, if you would filter by "product X" and "checked-in on list Y", the email might have been sent *even if those two properties are true for two different positions on the order, not in combination*. To make things worse, this was handled correctly only when sending only to atteendee email addresses, but the number of matching orders in the success message was off in this case. I fixed this by introducing a subquery for the postions.
